File: shared/otel-instrumentation.py
"""
Shared OpenTelemetry instrumentation configuration for Python services
"""
import os
from opentelemetry import trace, metrics
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor
from opentelemetry.sdk.metrics import MeterProvider
from opentelemetry.sdk.metrics.export import PeriodicExportingMetricReader
from opentelemetry.sdk.resources import Resource
from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter
from opentelemetry.exporter.otlp.proto.grpc.metric_exporter import OTLPMetricExporter
from opentelemetry.semconv.resource import ResourceAttributes


def setup_telemetry(service_name: str, service_version: str = "1.0.0"):
    """
    Setup OpenTelemetry instrumentation for a Python service
    
    Args:
        service_name: Name of the service
        service_version: Version of the service
    """
    # Get OTLP endpoint from environment or use default
    otlp_endpoint = os.getenv("OTEL_EXPORTER_OTLP_ENDPOINT", "adot-collector:4317")
    
    # Create resource with service information
    resource = Resource.create({
        ResourceAttributes.SERVICE_NAME: service_name,
        ResourceAttributes.SERVICE_VERSION: service_version,
        ResourceAttributes.SERVICE_NAMESPACE: "petshop-demo",
        ResourceAttributes.DEPLOYMENT_ENVIRONMENT: os.getenv("ENVIRONMENT", "demo"),
    })
    
    # Setup tracing
    trace_provider = TracerProvider(resource=resource)
    otlp_trace_exporter = OTLPSpanExporter(endpoint=otlp_endpoint, insecure=True)
    trace_provider.add_span_processor(BatchSpanProcessor(otlp_trace_exporter))
    trace.set_tracer_provider(trace_provider)
    
    # Setup metrics
    otlp_metric_exporter = OTLPMetricExporter(endpoint=otlp_endpoint, insecure=True)
    metric_reader = PeriodicExportingMetricReader(otlp_metric_exporter, export_interval_millis=10000)
    meter_provider = MeterProvider(resource=resource, metric_readers=[metric_reader])
    metrics.set_meter_provider(meter_provider)
    
    return trace.get_tracer(__name__), metrics.get_meter(__name__)

File: shared/otel-instrumentation.js
/**
 * Shared OpenTelemetry instrumentation configuration for Node.js services
 */
const { NodeSDK } = require('@opentelemetry/sdk-node');
const { OTLPTraceExporter } = require('@opentelemetry/exporter-trace-otlp-grpc');
const { OTLPMetricExporter } = require('@opentelemetry/exporter-metrics-otlp-grpc');
const { PeriodicExportingMetricReader } = require('@opentelemetry/sdk-metrics');
const { Resource } = require('@opentelemetry/resources');
const { SemanticResourceAttributes } = require('@opentelemetry/semantic-conventions');
const { HttpInstrumentation } = require('@opentelemetry/instrumentation-http');
const { ExpressInstrumentation } = require('@opentelemetry/instrumentation-express');

/**
 * Setup OpenTelemetry instrumentation for a Node.js service
 * 
 * @param {string} serviceName - Name of the service
 * @param {string} serviceVersion - Version of the service
 * @param {Array} additionalInstrumentations - Additional instrumentations to include
 * @returns {NodeSDK} Configured OpenTelemetry SDK instance
 */
function setupTelemetry(serviceName, serviceVersion = '1.0.0', additionalInstrumentations = []) {
  // Get OTLP endpoint from environment or use default
  const otlpEndpoint = process.env.OTEL_EXPORTER_OTLP_ENDPOINT || 'adot-collector:4317';
  
  // Create resource with service information
  const resource = new Resource({
    [SemanticResourceAttributes.SERVICE_NAME]: serviceName,
    [SemanticResourceAttributes.SERVICE_VERSION]: serviceVersion,
    [SemanticResourceAttributes.SERVICE_NAMESPACE]: 'petshop-demo',
    [SemanticResourceAttributes.DEPLOYMENT_ENVIRONMENT]: process.env.ENVIRONMENT || 'demo',
  });
  
  // Configure trace exporter
  const traceExporter = new OTLPTraceExporter({
    url: otlpEndpoint,
  });
  
  // Configure metric exporter
  const metricExporter = new OTLPMetricExporter({
    url: otlpEndpoint,
  });
  
  const metricReader = new PeriodicExportingMetricReader({
    exporter: metricExporter,
    exportIntervalMillis: 10000,
  });
  
  // Create SDK with instrumentations
  const sdk = new NodeSDK({
    resource,
    traceExporter,
    metricReader,
    instrumentations: [
      new HttpInstrumentation(),
      new ExpressInstrumentation(),
      ...additionalInstrumentations,
    ],
  });
  
  // Start the SDK
  sdk.start();
  
  // Graceful shutdown
  process.on('SIGTERM', () => {
    sdk.shutdown()
      .then(() => console.log('OpenTelemetry SDK shut down successfully'))
      .catch((error) => console.error('Error shutting down OpenTelemetry SDK', error))
      .finally(() => process.exit(0));
  });
  
  return sdk;
}

module.exports = { setupTelemetry };
